// file-tree-common/lib/createAction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var actions = {
  initialState: function initialState(state, rootPath) {
    return {
      type: 'initialState',
      payload: { state: state, rootPath: rootPath }
    };
  },
  change: function change(state) {
    return {
      type: 'change',
      payload: state
    };
  },
  event: function event(name, path, stat) {
    return {
      type: 'event',
      payload: { name: name, path: path, stat: stat }
    };
  },
  batch: function batch(actions) {
    return {
      type: 'batch',
      payload: actions
    };
  },
  response: function response(id, err, data) {
    return {
      type: 'response',
      meta: { id: id },
      error: !!err,
      payload: err ? err : data
    };
  },
  request: function request(id, methodName, args) {
    return {
      type: 'request',
      meta: { id: id },
      payload: { methodName: methodName, args: args }
    };
  },
  watchPath: function watchPath(path) {
    return {
      type: 'watchPath',
      payload: { path: path }
    };
  },
  rootPath: function rootPath(path, reset) {
    return {
      type: 'rootPath',
      payload: { path: path, reset: reset }
    };
  }
};

exports.default = function (type) {
  for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }

  return actions[type].apply(actions, args);
};
